browser/search_engines/search_engine_tracker.cc

```diff
@@ -0,0 +1,103 @@
+/* Copyright (c) 2020 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/browser/search_engines/search_engine_tracker.h"
+
+#include "base/metrics/histogram_macros.h"
+#include "base/strings/string_util.h"
+#include "chrome/browser/profiles/profile.h"
+#include "chrome/browser/search_engines/template_url_service_factory.h"
+#include "components/keyed_service/content/browser_context_dependency_manager.h"
+
+namespace {
+
+// Deduces the search engine from |type|, if nothing is found - from |url|.
+// Not all engines added by Brave are present in |SearchEngineType| enumeration.
+void RecordSearchEngineP3A(const GURL& search_engine_url,
+                           SearchEngineType type) {
+  SearchEngineP3A answer = SearchEngineP3A::kOther;
+
+  if (type == SEARCH_ENGINE_GOOGLE) {
+    answer = SearchEngineP3A::kGoogle;
+  } else if (type == SEARCH_ENGINE_DUCKDUCKGO) {
+    answer = SearchEngineP3A::kDuckDuckGo;
+  } else if (type == SEARCH_ENGINE_BING) {
+    answer = SearchEngineP3A::kBing;
+  } else if (type == SEARCH_ENGINE_QWANT) {
+    answer = SearchEngineP3A::kQwant;
+  } else if (type == SEARCH_ENGINE_OTHER) {
+    if (base::EndsWith(search_engine_url.host(), "startpage.com",
+                       base::CompareCase::INSENSITIVE_ASCII)) {
+      answer = SearchEngineP3A::kStartpage;
+    }
+  }
+
+  UMA_HISTOGRAM_ENUMERATION("Brave.Search.DefaultEngine", answer);
+}
+
+}  // namespace
+
+// static
+SearchEngineTrackerFactory* SearchEngineTrackerFactory::GetInstance() {
+  return base::Singleton<SearchEngineTrackerFactory>::get();
+}
+
+SearchEngineTrackerFactory::SearchEngineTrackerFactory()
+    : BrowserContextKeyedServiceFactory(
+          "SearchEngineTracker",
+          BrowserContextDependencyManager::GetInstance()) {
+  DependsOn(TemplateURLServiceFactory::GetInstance());
+}
+
+SearchEngineTrackerFactory::~SearchEngineTrackerFactory() {}
+
+KeyedService* SearchEngineTrackerFactory::BuildServiceInstanceFor(
+    content::BrowserContext* context) const {
```

Comment:
How about return nullptr if |context| is non normal profile?
I think this only tracks for normal profile?

Reply:
It already works like this, as per `BrowserContextKeyedServiceFactory::GetBrowserContextToUse`

Reply:
Oh, it's default behavior! :)